### PR TITLE
Moved to faster azure client, + restricted allowed namespaces characters...

### PR DESCRIPTION
**Pull request by jonasfj**
Fixing bug 1150304

Should enable us to add API end-points that redirects to artifacts directly from index... Or well, redirects to queue that redirects to artifacts :)

### CODE REVIEW

File: routes/api/v1.js [hunk 1]
```diff
@@ -59,6 +59,10 @@ var api = new base.API({
     "indexing tasks to make it easy to find artifacts, consider using the",
     "expiration date that the artifacts is assigned.",
     "",
+    "**Valid Characeters**, all keys in a namespace `<key1>.<key2>` must be",
```
Comment (gregarndt) [inline routes/api/v1.js]: Characters

File: index/handlers.js [hunk 2]
```diff
@@ -108,6 +108,8 @@ Handlers.prototype.completed = function(message) {
     return that.routeRegexp.test(route);
   }).map(function(route) {
     return that.routeRegexp.exec(route)[1];
+  }).filter(function(namespace) {
```
Comment (gregarndt) [inline index/handlers.js]: Is it possible to log somewhere when indices specified in the task will not be used?  If an invalid index route is provided, and the task isn't inserted there, maybe it wouldn't be immediately obvious to the user.  Not sure if it really matters much...

File: index/handlers.js [hunk 3]
```diff
@@ -108,6 +108,8 @@ Handlers.prototype.completed = function(message) {
...
+    return helpers.isValidNamespace(namespace);
```
Comment (gregarndt) [inline index/handlers.js]: So, this should provide us a list of namespaces that insertTask shouldn't choke on and return 400, but do we inform the user at all that some of their index routes might not be used while others might be?

**Comment by gregarndt** [comment]
This looks good, just had a couple of questions.

1. Is it intentional that we're ignoring invalid namespaces when a task is marked as completed? I would assume so because at that point there's really no way to reject anything because the task is already completed.  Just noticed the difference in behavior when a task is marked as completed and the handler does something with it vs the client calling the insertTask endpoint (which will return a 400).
2. How are client api changes like this tested?  Is it the responsibility of the taskcluster-test tasks to test these api changes to make sure it handles a 400 being returned?  As of now the clients expect things to succeed for most namespaces but now they will need to make sure this exception is caught if they are calling it directly.

File: routes/api/v1.js (hunk 1 above)
Comment (jonasfj) [inline routes/api/v1.js]: There is no excuse...  my editor even has spellchecker activated....

**Comment by jonasfj** [comment]
> Is it intentional that we're ignoring invalid namespaces when a task is marked as completed? I would assume so because at that point there's really no way to reject anything because the task is already completed. Just noticed the difference in behavior when a task is marked as completed and the handler does something with it vs the client calling the insertTask endpoint (which will return a 400).

Yeah, as you know we've discussed the option of some sort of "note" feature for reporting errors in handlers... IMO, it's probably best if people use the insertTask end-point  and fail the task if that fails...
Certainly the sane thing to do if something relies on the indexed value.

I don't see a good way of reporting errors from handlers... We don't really want tasks to fail later. Pulse handlers deals with events, I don't see a good pattern for reporting errors from distributed event handlers. The "note" feature wouldn't even really solve the problem.

Note, as we discussed in the "note" feature bug... We could insert a message in "note.<taskId>.<component>" with the error... So if we get an invalid namespace from a task in the handler, we could insert "note.<taskId>.index" and set the data property to a JSON message explaining what is wrong...
At least it would be possible to see the error without going through the index.taskcluster.net log...

> How are client api changes like this tested? Is it the responsibility of the taskcluster-test tasks to test these api changes to make sure it handles a 400 being returned? As of now the clients expect things to succeed for most namespaces but now they will need to make sure this exception is caught if they are calling it directly.

Existing clients inserts tasks under namespaces that conforms the this logic... So they shouldn't be affected... at least I couldn't find any that used namespaces with illegal characters...
And yes, if you use index.insertTask inside a taskcluster task, it's your responsibility to exit non-zero if the request fails...

**Pull request merged**